### PR TITLE
Add global user prompt prefix feature with cache partitioning

### PR DESCRIPTION
**Pull request by danielaskdd**
## Description

This PR implements a server-side global user prompt prefix feature that allows operators to prepend instructions to every request's `user_prompt`. The prefix is composed with the request's user prompt and reaches the LLM, while being carefully integrated into the answer cache key to ensure cache partitioning when the prefix changes.

## Related Issues

N/A

### Core Feature Implementation
- **`lightrag/utils.py`**: Added `EffectiveUserPrompt` NamedTuple and `resolve_user_prompt()` function to compose server-side prefix with request user_prompt via byte concatenation (no separator injected)
- **`lightrag/prompt.py`**: Added `load_user_prompt_prefix_source()` to load prefix from `USER_PROMPT_PREFIX` env var or `USER_PROMPT_PREFIX_FILE` (`.md`/`.txt` files in `PROMPT_DIR/user_prompt`). Refactored prompt file resolution into shared `_resolve_prompt_file()` helper
- **`lightrag/lightrag.py`**: Added `user_prompt_prefix` parameter to `LightRAG` class with default factory loading from environment
- **`lightrag/base.py`**: Added `disable_user_prompt_prefix: bool = False` field to `QueryParam` to allow per-request opt-out
- **`lightrag/operate.py`**: Integrated prefix composition into `kg_query()` and `naive_query()` by using composed text in answer cache key (keeping `_ANSWER_CACHE_POLICY_VERSION` at v2 for backward compatibility)

### API Integration
- **`lightrag/api/routers/query_routes.py`**: Added `disable_user_prompt_prefix` field to `QueryRequest` with proper documentation
- **Environment files**: Updated `env.example` and `env.docker-compose-full` with `USER_PROMPT_PREFIX` and `USER_PROMPT_PREFIX_FILE` documentation
- **Sample file**: Added `prompts/samples/user_prompt_prefix.sample.md` as reference

### WebUI Integration
- **`lightrag_webui/src/components/retrieval/QuerySettings.tsx`**: Added UI toggle for `disable_user_prompt_prefix`
- **`lightrag_webui/src/stores/querySettings.ts`**: Added `disable_user_prompt_prefix: false` to default query settings
- **`lightrag_webui/src/api/lightrag.ts`**: Updated TypeScript types for `QueryRequest`
- **`lightrag_webui/src/features/retrieval/serializeQueryRequest.test.ts`**: Added tests for prefix serialization
- **Localization files**: Updated all locale JSON files (en, zh, ja, ko, fr, de, ru, uk, ar, vi, zh_TW) with new UI strings

### Documentation
- **`docs/ProgramingWithCore.md`**: Added section explaining global user prompt prefix feature with code examples
- **`README.md`, `README-ja.md`, `README-zh.md`**: Updated environment variable documentation

### Testing
- **`tests/llm/test_query_cache_user_prompt_prefix.py`**: Comprehensive test suite (482 lines) covering:
  - Prefix reaches the model and is prepended to system prompt
  - Cache partitioning when prefix changes
  - Cache hits with same prefix
  - Boundary conditions between prefix and user_prompt
  - `disable_user_prompt_prefix` opt-out behavior
  - Backward compatibility with pre-prefix cache entries
  - Cache metadata transparency
  - Token budget accounting for prefix length
  
- **`tests/test_user_prompt_prefix_env.py`**: Environment variable loading tests (203 lines) covering:
  - Inline `USER_PROMPT_PREFIX` env var handling
  - File-based `USER_PROMPT_PREFIX_FILE` loading
  - Multiline file content preservation
  - Failure modes (missing files, invalid names, encoding errors) degrade gracefully
  
- **`tests/utils/test_resolve_user_prompt.py`**: Unit tests for prompt composition (144 lines) covering:
  - Composition order and no separator injection

https://claude.ai/code/session_01WYAiVnoKKS2W3ufYV9rdn2

### CODE REVIEW

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/prompt.py [hunk 1]
```diff
@@ -583,16 +617,88 @@ def resolve_entity_type_prompt_path(prompt_file_name: str | Path) -> Path:
...
+        logger.warning(
+            "Both USER_PROMPT_PREFIX_FILE and USER_PROMPT_PREFIX are set; "
+            "using the file '%s' and ignoring the inline value.",
+            file_name,
         )
 
-    return get_entity_type_prompt_dir() / candidate.name
+    try:
+        prefix_path = resolve_user_prompt_prefix_path(file_name)
+    except ValueError as exc:
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/prompt.py]: Valid finding, fixed in a8cf2c5 — and the impact is worse than described.

It is not that the prefix is skipped: because the loader runs as a dataclass `default_factory`, the escaping exception makes `LightRAG()` unconstructible entirely, with a traceback far from the environment variable that caused it. That directly contradicts the function's own documented contract.

It also does not need a symlink loop to reproduce. `Path.expanduser()` raises `RuntimeError` whenever a home directory cannot be determined, so a `PROMPT_DIR` naming an unresolvable home is enough:

```
$ PROMPT_DIR='~nosuchuser12345/prompts' USER_PROMPT_PREFIX_FILE='house.md' python -c "..."
RuntimeError: Could not determine home directory.
```

The except is now `(ValueError, OSError, RuntimeError)` — `ValueError` for the sandbox's own rejections, `RuntimeError` for `expanduser()` and for `resolve()` symlink loops on Python <= 3.12, `OSError` for `resolve()` filesystem errors. Two regression tests were added, both verified to fail without the fix: one asserts the loader returns `""` and logs, the other asserts the `default_factory` itself does not raise, which is the consequence that actually matters. They use the `~unknownuser` form rather than a symlink loop since it needs no filesystem setup and behaves the same across platforms.

`resolve_entity_type_prompt_path` is deliberately unchanged. It shares the same `_resolve_prompt_file` sandbox but has the opposite contract — a misconfigured `ENTITY_TYPE_PROMPT_FILE` raises so startup fails loudly — so only the prefix loader, which promises to degrade, widens its except.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/base.py [hunk 2]
```diff
@@ -150,6 +150,19 @@ class QueryParam:
...
     Its purpose is to let the user customize the way LLM generates the response.
+
+    The server may prepend a global prefix (``USER_PROMPT_PREFIX`` /
+    ``LightRAG.user_prompt_prefix``) to this value; set
+    ``disable_user_prompt_prefix`` to opt out. Note that ``bypass`` mode ignores
+    this field entirely, and a caller-supplied ``system_prompt`` without a
+    ``{user_prompt}`` placeholder silently drops it.
+    """
+
+    disable_user_prompt_prefix: bool = False
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/base.py]: Valid, fixed in 720d64e. Confirmed `QueryParam` is a plain `@dataclass` with no `kw_only=True`, so the field order is public API and the break is real and silent:

```
# before the fix, 15th positional False:
disable_user_prompt_prefix = False   # bound here
enable_rerank = True                 # kept its RERANK_BY_DEFAULT value
```

`disable_user_prompt_prefix` now sits after `include_references`, the last field, which restores the original signature exactly. A comment on the field records why it is declared there so the next addition does not repeat this.

Added `tests/test_queryparam_positional_compatibility.py`, which pins the legacy positional prefix as a literal list and asserts the two concrete bindings (15th → `enable_rerank`, 16th → `include_references`). Verified it fails — all four cases — when the field is moved back.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lightrag/operate.py [hunk 3]
```diff
@@ -5696,7 +5720,7 @@ async def _build_context_str(
     pre_sys_prompt = sys_prompt_template.format(
         context_data="",  # Empty for overhead calculation
         response_type=response_type,
-        user_prompt=user_prompt,
+        user_prompt=effective_user_prompt.text,
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/operate.py:5741]: Valid, fixed in 720d64e by forwarding the system prompt, as suggested. Investigating it turned up a defect older than this PR.

`_build_context_str` read `global_config["system_prompt_template"]` — and **nothing anywhere in the repo writes that key** (one grep hit: the read itself). So the estimate always used `PROMPTS["rag_response"]`, never the caller's template. `kg_query` selects its template at the `sys_prompt_temp = system_prompt if system_prompt else ...` line, which runs *after* `_build_query_context` has already budgeted the context. So a custom `system_prompt` was budgeted at the wrong size regardless of this PR — a long one risks overfilling the context. Putting the prefix into that estimate is what made the pre-existing gap start costing chunks. `naive_query` computes its budget inline in the function that received `system_prompt` and has always been correct; that asymmetry is the tell.

`system_prompt` is now threaded `kg_query` → `_build_query_context` → `_build_context_str` (both new parameters are keyword-with-default and appended, so the one positional test call and the two `(*args, **kwargs)` stubs are unaffected). Precedence is `system_prompt` arg > `global_config["system_prompt_template"]` > `PROMPTS["rag_response"]` — the unwritten key is kept as a lower-priority fallback rather than deleted, since it is undocumented but a downstream user could be setting it on their own config dict.

Measured, with a 1-char-per-token tokenizer and a 3000-token prefix:

| template | budget without prefix | with prefix |
|---|---|---|
| default (has `{user_prompt}`) | 26438 | 23438 |
| custom, no placeholder | 29205 | 29205 |

The prefix is charged only where it is actually rendered, and the custom template is now budgeted at its own size. Test verified to fail without the forwarding.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/lightrag.py [hunk 4]
```diff
@@ -463,6 +464,18 @@ class LightRAG(_RoleLLMMixin, _StorageMigrationMixin, _PipelineMixin):
     )
     """Append each chunk's parent heading path as a `content_headings` field in the chunk JSON sent to the LLM."""
 
+    user_prompt_prefix: str = field(default_factory=load_user_prompt_prefix_source)
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/lightrag.py]: Valid, fixed in 80c72ae. This is the same defect as the `QueryParam` one earlier in this PR, in the other dataclass the feature touched — I fixed the first and did not check the second.

The blast radius is much larger here:

```
total init params: 72
user_prompt_prefix was at position 17
parameters shifted: 55        (QueryParam shifted 2)
```

The failure mode you describe is right, and the delayed-error part is what makes it nasty: an int meant for `entity_extract_max_gleaning` binds to a string field and only raises during prompt composition, while the other 54 values silently configure the wrong options.

The field now sits after `_storages_status`, the last pre-existing init field — the only position that restores the signature exactly. Verified: `entity_extract_max_gleaning` is back at index 16, `user_prompt_prefix` is last at 71.

`tests/test_queryparam_positional_compatibility.py` is renamed to `test_dataclass_positional_compatibility.py` and now covers both classes. For `LightRAG` it pins the legacy 71-parameter order as a literal, since that is the only form that catches an insertion anywhere in the middle; it introspects `inspect.signature` rather than constructing `LightRAG`, whose `__post_init__` does real storage setup. All three new assertions were verified to fail with the field moved back.

I also swept the rest of this PR's surface for the same class of break: `QueryRequest` is a pydantic model (keyword-only), and the two parameters added to `_build_context_str` / `_build_query_context` are appended keyword-with-defaults. `LightRAG` was the last exposure.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/operate.py [hunk 5]
```diff
@@ -6605,7 +6647,7 @@ async def naive_query(
     # Create a preliminary system prompt with empty content_data to calculate overhead
     pre_sys_prompt = sys_prompt_template.format(
         response_type=response_type,
-        user_prompt=user_prompt,
+        user_prompt=effective_user_prompt.slot,
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/operate.py:6659]: Declining this one — the charge is deliberate and the behaviour is unchanged in 81037a1. The observation is accurate (the prompt really is never rendered on that path), but the conclusion inverts what these switches are for.

`only_need_context` and `only_need_prompt` are debug switches whose purpose is to **preview the real request**, not to maximise recall. If retrieval-only calls skipped the charge, they would report more chunks than a live query retrieves — so an operator who inspected the context, judged it sufficient, and then ran the query for real would be silently truncated. The debug switch would become the source of the very problem it exists to expose. Reporting a smaller-but-accurate context is the correct trade.

Measured on the KG path with a 3000-token prefix:

```
only_need_context=False -> chunk_token_limit = 23426
only_need_context=True  -> chunk_token_limit = 23426
```

That equality is the property worth having, and it now has a test: `test_only_need_context_budgets_exactly_like_a_real_query`, verified to fail if the suggested skip is applied.

On the `ProgramingWithCore.md:616` contract — the citation is fair, the sentence was too broad, so it is clarified rather than the code changed to match the reading. "Does not participate in retrieval" means `user_prompt` does not influence *what* is found or *how* it is ranked (no keyword extraction, no vector search, no rerank). It does consume token budget, because it genuinely occupies space in the final prompt alongside the context.

Separately confirmed while investigating, since the two are easy to conflate: `only_need_prompt` does return the composed prefix. Both paths render `sys_prompt` from `effective_user_prompt.slot` and hand back exactly that:

```
with user_prompt   -> ['6. Additional Instructions: ', 'Answer in Chinese.', 'Reply in one sentence.']
empty user_prompt  -> ['6. Additional Instructions: ', 'Answer in Chinese.']
```

Both modes, both cases, now covered by a test — including the empty-`user_prompt` case on this exit, which was previously only asserted on the LLM-bound prompt.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review
